idxmgmt/supporter.go

```diff
@@ -71,48 +88,51 @@ type ilmIndexSelector struct {
 	st           *indexState
 }
 
-func newSupporter(
-	log *logp.Logger,
-	info beat.Info,
-	templateConfig template.TemplateConfig,
-	ilmConfig ilm.Config,
-	outConfig common.ConfigNamespace,
-) (*supporter, error) {
+func newSupporter(log *logp.Logger, info beat.Info, cfg *IndexManagementConfig) (*supporter, error) {
 
 	var (
 		unmanagedIdxCfg unmanaged.Config
-		mode            = ilmConfig.Mode
+		mode            = cfg.ILM.Mode
 		st              = indexState{}
 	)
 
-	if outConfig.Name() == esKey {
-		if err := outConfig.Config().Unpack(&unmanagedIdxCfg); err != nil {
+	if cfg.Output.Name() == esKey {
+		if err := cfg.Output.Config().Unpack(&unmanagedIdxCfg); err != nil {
 			return nil, fmt.Errorf("unpacking output elasticsearch index config fails: %+v", err)
 		}
 
-		if err := checkTemplateESSettings(templateConfig, &unmanagedIdxCfg); err != nil {
+		if err := checkTemplateESSettings(cfg.Template, &unmanagedIdxCfg); err != nil {
 			return nil, err
 		}
 	}
 
-	if outConfig.Name() != esKey ||
-		ilmConfig.Mode == libilm.ModeDisabled ||
-		ilmConfig.Mode == libilm.ModeAuto && unmanagedIdxCfg.Customized() {
-
+	var disableILM bool
+	if cfg.Output.Name() != esKey || cfg.ILM.Mode == libilm.ModeDisabled {
+		disableILM = true
+	} else if cfg.ILM.Mode == libilm.ModeAuto {
+		// ILM is set to "auto": disable if we're using data streams,
+		// or if we're not using data streams but we're using customised,
+		// unmanaged indices.
+		if cfg.DataStreams || unmanagedIdxCfg.Customized() {
+			disableILM = true
+		}
+	}
```

Comment:
This indicates that when `apm-server.ilm.enabled=true` and `apm-server.data_streams.enabled=true` the ILM config is stronger than the data streams. While I understand that data streams is early stage and not enabled by default, it is the new strategy so would flip this and make data streams overrule ILM if configured. 

Looking at https://github.com/elastic/apm-server/pull/4409/files#diff-257c35f49365729cf758ca1de95ac8957ae2b337b0bd85ab21e7a992a925e94bR168 this also doesn't seem aligned with the `BuildSelector` that is returned - as soon as `cfg.DataStream` is true it will return the data stream supporter. 
So what I could observe is that if `apm-server.ilm.enabled=true` and `apm-server.data_streams.enabled=true` data are indexed to the new data stream indices, but the ILM indices are still created.

Reply:
This code here is only meant to be about changing the default behaviour, when the user hasn't set anything in config. I was planning to create a follow up PR with additional config validation that would prevent setting `apm-server.ilm.*`, `output.index`, and `output.indices` when `apm-serverdata_streams.enabled=true`. Would that address your concern, and are you OK with it coming in a follow up?

Reply:
That sounds totally fine.